sematic/graph.py

```diff
@@ -0,0 +1,463 @@
+                downstream_ids.append(output_edge.destination_run_id)
+                downstream_ids += self._get_run_downstream_ids(
+                    output_edge.destination_run_id
+                )
+
+        return list(set(downstream_ids))
+
+    def clone_futures(
+        self, storage: Storage, reset_from: Optional[str] = None
+    ) -> Tuple[
+        OrderedDict[str, AbstractFuture],
+        Dict[str, Dict[str, Artifact]],
+        Dict[str, Artifact],
+    ]:
+        """
+        Clones the current graph into new futures.
+
+        Future state is set as follows:
+
+        - If run is RAN or RESOLVED, future state is set accordingly. If
+          reset_from, see behavior below.
+
+        - If run is FAILED, NESTED_FAILED, or CANCELED, future state is set to
+          CREATED
+
+        If reset_from is not None, all ancestor runs are set to RAN or CREATED
+        depending on original run status. Cloned futures of all downstream runs
+        of reset_from are set to CREATED. All descendant runs of all downstream
+        runs of all ancestors are set to CREATED.
+
+        If reset_from is None, only FAILED, NESTED_FAILED, and CANCELED runs
+        will be reset.
+
+        Parameters
+        ----------
+        storage: Storage
+            The storage class to retrieve artifact values and set future.value
+            and future.kwargs appropriately.
+
+        reset_from: Optional[str]
+            Force reset other runs than only failed ones.
+
+        Returns
+        -------
+        Tuple[List[AbstractFuture], Dict[str, Dict[str, Artifact]], Dict[str,
+        Artifact]]
+            A tuple whose first element is a list of cloned futures, grouped by
+            nested layers (outermost first), and sorted by reverse execution
+            order within each layer. The second element is
+
+        """
+        value_by_artifact_id: Dict[str, Any] = {}
+
+        futures_by_original_id: Dict[str, AbstractFuture] = {}
+        input_artifacts: Dict[str, Dict[str, Artifact]] = defaultdict(dict)
+        output_artifacts: Dict[str, Artifact] = {}
+
+        # We skip descendants of all downstream of reset point
+        # plus descendants of downstreams of ancestors
+        # The skipped futures will be naturally re-created by the new graph
+        # resolution
+        skip_run_ids: List[str] = []
+
+        # reset = forcing future state to CREATED or RAN
+        # Considering reset_from and ancestors runs, we reset the run and
+        # all downstream
+        reset_run_ids: List[str] = []
+
+        if reset_from is not None:
+            reset_run_ids.append(reset_from)
+            ancestor_run_ids = self._get_run_ancestor_ids(reset_from)
+            reset_run_ids += ancestor_run_ids
+
+            for ancestor_run_id in [reset_from] + ancestor_run_ids:
+                downstream_run_ids = self._get_run_downstream_ids(ancestor_run_id)
+                reset_run_ids += downstream_run_ids
```

Comment:
Shouldn't downstream runs of the reset run be skipped instead of reset?
We want to mainly cover the use case where there's a corner-case bug in the code, and allow the user to make changes. Those changes could impact the signature of its output artifacts, so the previous downstream runs would need to be re-created.

Reply:
No, all futures in a given layer are known at once from the execution of the parent future. Therefore, downstream futures should be there, but reset.
For any future that re-executes (all reset futures), the new code will be used. Resetting just means resetting the future's state so that it gets re-executed. But the code will always be the new code since we are doing `run.get_func()` which imports the function from the current code.

Reply:
Please add this explanation as a code comment here.